### PR TITLE
[CDAP-17787] Add auth enforcement for preference service

### PR DESCRIPTION
**Pull request by wyzhang**

### CODE REVIEW

**Review by CuriousVini — APPROVED** [review]

File: cdap-app-fabric/src/main/java/io/cdap/cdap/config/PreferencesService.java [hunk 1]
```diff
@@ -101,10 +115,9 @@ private void setConfig(EntityId entityId,
   /**
    * Validate the profile status is enabled and set the preferences
    */
-  private void setConfig(ProfileStore profileStore, PreferencesTable preferencesTable, EntityId entityId,
-                         Map<String, String> propertyMap)
-    throws NotFoundException, ProfileConflictException, BadRequestException, IOException {
-
+  private void setConfig(ProfileStore profileStore, PreferencesTable preferencesTable,
+                         @Name("entityId") EntityId entityId, Map<String, String> propertyMap) throws Exception {
```
Comment (dli357) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/config/PreferencesService.java]: nit: remove Name annotation

Comment (yaojiefeng) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/config/PreferencesService.java]: Why do we need a name annotation on this?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/config/PreferencesService.java [hunk 2]
```diff
@@ -101,10 +115,9 @@ private void setConfig(EntityId entityId,
...
+    authorizationEnforcer.enforce(entityId, authenticationContext.getPrincipal(), Action.EXECUTE);
     boolean isInstanceLevel = entityId.getEntityType().equals(EntityType.INSTANCE);
     NamespaceId namespaceId = isInstanceLevel ?
       NamespaceId.SYSTEM : ((NamespacedEntityId) entityId).getNamespaceId();
```
Comment (dli357) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/config/PreferencesService.java]: It looks like there is a distinction which always attempts to enforce on a namespace entity. Should we performing the enforcement at the lower level in this case?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/config/PreferencesService.java [hunk 3]
```diff
@@ -101,10 +115,9 @@ private void setConfig(EntityId entityId,
   /**
    * Validate the profile status is enabled and set the preferences
    */
-  private void setConfig(ProfileStore profileStore, PreferencesTable preferencesTable, EntityId entityId,
-                         Map<String, String> propertyMap)
-    throws NotFoundException, ProfileConflictException, BadRequestException, IOException {
-
+  private void setConfig(ProfileStore profileStore, PreferencesTable preferencesTable,
+                         @Name("entityId") EntityId entityId, Map<String, String> propertyMap) throws Exception {
+    authorizationEnforcer.enforce(entityId, authenticationContext.getPrincipal(), Action.EXECUTE);
```
Comment (yaojiefeng) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/config/PreferencesService.java]: here entity id can be InstanceId(""), a hardcoded value for instance level preferences, will user have permission on this?

Comment (yaojiefeng) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/config/PreferencesService.java]: Also, should he action be ADMIN or EXECUTE? what is the scenario we use ADMIN and when we use EXECUTE

File: cdap-app-fabric/src/main/java/io/cdap/cdap/config/PreferencesService.java [hunk 4]
```diff
@@ -155,7 +168,8 @@ private void setConfig(ProfileStore profileStore, PreferencesTable preferencesTa
     }
   }
 
-  private void deleteConfig(EntityId entityId) {
+  private void deleteConfig(EntityId entityId) throws Exception {
+    authorizationEnforcer.enforce(entityId, authenticationContext.getPrincipal(), Action.EXECUTE);
     TransactionRunners.run(transactionRunner, context -> {
       PreferencesTable dataset = new PreferencesTable(context);
       Map<String, String> oldProp = dataset.getPreferences(entityId).getProperties();
```
Comment (dli357) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/config/PreferencesService.java]: Same comment as above regarding moving the enforcement check downwards.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/config/PreferencesService.java [hunk 5]
```diff
@@ -155,7 +168,8 @@ private void setConfig(ProfileStore profileStore, PreferencesTable preferencesTa
     }
   }
 
-  private void deleteConfig(EntityId entityId) {
+  private void deleteConfig(EntityId entityId) throws Exception {
+    authorizationEnforcer.enforce(entityId, authenticationContext.getPrincipal(), Action.EXECUTE);
```
Comment (yaojiefeng) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/config/PreferencesService.java]: Same here, how does InstanceId work?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/config/PreferencesService.java [hunk 6]
```diff
@@ -53,35 +59,42 @@
...
+    this.authorizationEnforcer = authorizationEnforcer;
+    this.authenticationContext = authenticationContext;
     MultiThreadMessagingContext messagingContext = new MultiThreadMessagingContext(messagingService);
     this.adminEventPublisher = new AdminEventPublisher(cConf, messagingContext);
     this.transactionRunner = transactionRunner;
   }
 
-  private PreferencesDetail get(EntityId entityId) {
+  private PreferencesDetail get(EntityId entityId) throws Exception {
+    AuthorizationUtil.ensureAccess(entityId, authorizationEnforcer, authenticationContext.getPrincipal());
```
Comment (yaojiefeng) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/config/PreferencesService.java]: same here

File: cdap-app-fabric/src/main/java/io/cdap/cdap/config/PreferencesService.java [hunk 7]
```diff
@@ -53,35 +59,42 @@
...
     return TransactionRunners.run(transactionRunner, context -> {
       return new PreferencesTable(context).getPreferences(entityId);
     });
   }
 
-  private PreferencesDetail getResolved(EntityId entityId) {
+  private PreferencesDetail getResolved(EntityId entityId) throws Exception {
+    AuthorizationUtil.ensureAccess(entityId, authorizationEnforcer, authenticationContext.getPrincipal());
```
Comment (yaojiefeng) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/config/PreferencesService.java]: same here

File: cdap-app-fabric/src/main/java/io/cdap/cdap/config/PreferencesService.java [hunk 8]
```diff
@@ -176,101 +190,118 @@ private void deleteConfig(EntityId entityId) {
...
+    throws Exception {
     return getResolved(programId);
   }
 
   /**
    * Set instance level preferences
    */
-  public void setProperties(Map<String, String> propMap)
-    throws NotFoundException, ProfileConflictException, BadRequestException {
+  public void setProperties(Map<String, String> propMap) throws Exception {
```
Comment (dli357) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/config/PreferencesService.java]: Same comment as above, seems like we support a level higher than namespace.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/config/PreferencesService.java (hunk 1 above)
Comment (wyzhang) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/config/PreferencesService.java]: Done

Comment (wyzhang) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/config/PreferencesService.java]: Failed to clean it up. Leftover from previous change.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/config/PreferencesService.java (hunk 6 above)
Comment (wyzhang) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/config/PreferencesService.java]: As discussed, auth extension will return if it is an instance id

File: cdap-app-fabric/src/main/java/io/cdap/cdap/config/PreferencesService.java (hunk 7 above)
Comment (wyzhang) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/config/PreferencesService.java]: same here

File: cdap-app-fabric/src/main/java/io/cdap/cdap/config/PreferencesService.java (hunk 2 above)
Comment (wyzhang) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/config/PreferencesService.java]: As discussed, auth extension can chose to ignore if it is an instance id

File: cdap-app-fabric/src/main/java/io/cdap/cdap/config/PreferencesService.java (hunk 4 above)
Comment (wyzhang) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/config/PreferencesService.java]: same here

**Review by dli357 — APPROVED** [review]
LGTM. Thanks!

File: cdap-app-fabric/src/main/java/io/cdap/cdap/config/PreferencesService.java (hunk 1 above)
Comment (wyzhang) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/config/PreferencesService.java]: Done

File: cdap-app-fabric/src/main/java/io/cdap/cdap/config/PreferencesService.java (hunk 3 above)
Comment (wyzhang) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/config/PreferencesService.java]: Fixed.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/config/PreferencesService.java (hunk 5 above)
Comment (wyzhang) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/config/PreferencesService.java]: As discussed, auth extension can choose to ignore or enforce if it is instance id

File: cdap-app-fabric/src/main/java/io/cdap/cdap/config/PreferencesService.java (hunk 8 above)
Comment (wyzhang) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/config/PreferencesService.java]: Same here. Auth extension can decide to ignore or enforce if it is an instance id

**Review by yaojiefeng — APPROVED** [review]

**Review by yaojiefeng — APPROVED** [review]

**Review by dli357 — APPROVED** [review]
LGTM

**Comment by wyzhang** [comment]
@dli357 @chtyim and I had some more discussions. We agreed on making auth enforcement call at the http handler for a few benefits 
1) Allow easy in-memory service to service communication
2) avoid throw generic Exception (since auth enforce just throws Exception causing all callers to just throw generic exception) 
3) code is simpler

**Review by CuriousVini — APPROVED** [review]

**Pull request merged**